### ISSUE TITLE
Wrong type inferred instead of using Any with structlog

### ISSUE THREAD

**Issue by sbdchd** [microsoft/pyright#799]
**Describe the bug**
I'm trying to setup pylance on an existing project but have run into some issues which I think are pyright related.

Specifically, the `bind` method in [structlog](https://www.structlog.org/en/stable/) is being incorrectly inferred instead of being typed as `Any`.

note: structlog doesn't have any type stubs in the package itself or in typeshed.

**To Reproduce**

```python
import structlog

logger = structlog.get_logger()

def main() -> None:
    log = logger.bind(user_id="foo")

    reveal_type(logger.bind)

    log.info("foobar")

main()
```

hover over `logger.bind(user_id="foo")` via pylance enabled in VSCode

```
Argument missing for parameter "self" Pylance
```

mypy:
```
❯ mypy --config-file mypy.ini .
main.py:8:17: note: Revealed type is 'Any'
```

pyright from the command line doesn't report any issues, so maybe this is already fixed but not available through pylance yet?

```
❯ pyright
No configuration file found.
stubPath /Users/steve/Downloads/structlog-pyright-bug/typings is not a valid directory.
Searching for source files
Auto-excluding /Users/steve/Downloads/structlog-pyright-bug/.venv
Found 1 source file
/Users/steve/Downloads/structlog-pyright-bug/main.py
  8:17 - warning: Type of "logger.bind" is "logger"
0 errors, 1 warning 
Completed in 0.715sec
```

If I update the code to declare logger with a type the problem goes away:

```python
logger: structlog.stdlib.BoundLogger = structlog.get_logger()
```

**Expected behavior**

Either defaulting to Any for dependencies or updating the inference

**VS Code extension or command-line**

Pylance version: `v2020.6.1`

Pyright command line version: `1.1.50`

mypy version: `0.782`

**Comment by erictraut** [comment]
By default, Pyright relies on type stubs to provide type definitions. If a type stub can't be found, the entire module is treated as `Any`. However, Pyright offers a mode (controlled by the pyright.useLibraryImplementations setting) whereby it will attempt to infer types based on the library implementation if a type stub isn't available. These inferred types tend to be incomplete because there are limits to static type inference, but they provide value if you're mainly interested in completion suggestions.

By default, the Pyright extension sets typeCheckingMode to "basic" and useLibraryImplementations to false. Pylance, by contrast, defaults typeCheckingMode to "off" and useLibraryImplementations to true. The assumption is that most Pylance users are more interested in completion suggestions than in type checking, whereas Pyright users are more interested in using type checking to find bugs in their code (at the expense of some false positive errors that need to be worked around).

If you pass in "--lib" to the command-line version of Pyright, it will enable the "useLibraryImplementations" setting, and you should see the same error that you do with the Pyright or Pylance extension.

I presume that you have pyright.useLibraryImplementations set to true (or are using Pylance, in which case it defaults to true). That means types will be inferred from the library implementation rather than simply assuming `Any`. 

In this particular case, the "structlog" library code is doing something tricky (and kind of gross). It is overriding a method with a callable instance variable. When a callable object is stored on a class instance, it is presumed to be "bound" to that object, so no "self" argument is passed to that callable. 

For example, func1 and func2 are both callable, but the implementation of func2 has a `self` parameter but the implementation of func1 does not. Yet, both are invoked in the same way.
```
def func1(a: int):
    print(a)

class Foo:
   def __init__(self):
       self.func1 = func1

   def func2(self, a: int):
        print(a)

foo = Foo()
foo.func1(2)
foo.func2(2)
```

Pyright's logic was confused by the fact that there were two versions of the `bind` method — one declared in the class (with a `self` parameter) and one later stored on the class instance (without a `self` parameter). I've updated the logic to deal with this particular case.

**Comment by erictraut** [comment]
This is addressed in Pyright 1.1.51, which I just published. It will be included in the next release of Pylance.
